### ISSUE TITLE
False positive TypeDoesNotContainType with literal-string and class-string

### ISSUE THREAD

**Issue by AndrolGenhald** [vimeo/psalm#8159]
hi there,

i would be happy if someone could let me know, how I can solve this:
https://psalm.dev/r/0547b7734b

thanks a lot !!

_Originally posted by @michabbb in https://github.com/vimeo/psalm/discussions/8156_

**Comment by someniatko** [comment]
you can workaround it like this https://psalm.dev/r/3e5d113867

**Comment by orklah** [comment]
For now, I don't think there is a single literal string that is accepted as a class-string. We could change that but I'm not sure I see the point. If all your variable are known in advance, why not just go ahead and use ::class notation.

On non literal strings, class_exists should be less annoying I think

**Comment by michabbb** [comment]
hi there, my problem is, that the class gets created by variables, so I still don´t see a solution here:

https://psalm.dev/r/ec09283594

**Comment by AndrolGenhald** [comment]
> For now, I don't think there is a single literal string that is accepted as a class-string.

Actually [it's fine](https://psalm.dev/r/7beecf322e) if you use a literal string directly, it just [doesn't correctly identify it as a class when using it elsewhere](https://psalm.dev/r/ad8cd01abe).

@michabbb `'\this\is\my\\'.$test.'\bla::class'` can't be a class anyway, so you probably meant `'this\is\my\\'.$test.'\bla'`, but you could just as easily use `\this\is\my\test\bla::class`.

**Comment by michabbb** [comment]
@AndrolGenhald sorry, i don´t get it. my class name is not a fixed string, it´s built by variables. can you show me an example like that where psalm is not complaining?

**Comment by AndrolGenhald** [comment]
@michabbb It's built by variables but at least in the examples you've given the variable will always be a known literal string, which could just as easily be replaced with a FQCN. Do you have an example that Psalm complains about that couldn't be replaced by a FQCN?

**Comment by michabbb** [comment]
@AndrolGenhald long day, maybe i am too stupid today, sorry for that 😕 I don´t really get you. can we please make it the other way around: can you show me an example where the class-name gets created dynamically?

**Comment by AndrolGenhald** [comment]
> can you show me an example where the class-name gets created dynamically?

https://psalm.dev/r/15e0553e14, https://psalm.dev/r/39e4bd0b54, https://psalm.dev/r/0c1e6896cd, etc

Psalm only complains when the variable you're using has a known literal value. If it's an unknown string it's fine.

> my class name is not a fixed string, it´s built by variables.

I think this is the misunderstanding, it's _technically_ not a fixed string, but it might as well be. Psalm is smart about concatenating known string literals and keeping track of them, so it knows that for `$test = "foo" . "bar"` `$test` will always be exactly `"foobar"` and `$test . "baz"` will be exactly `"foobarbaz"`.

**Comment by michabbb** [comment]
@AndrolGenhald in your example there is no existing namespace so I do myself hard to map your examples into my real-world code. can you make this work without any function?  --> https://psalm.dev/r/11458d0096
or is your/the solution "using an extra function" like "getReflectionClass" ??

**Comment by michabbb** [comment]
i am sure i am overseeing something here, but if you see my screenshot, you might understand why this is totally confusing me 🙈 
![image](https://user-images.githubusercontent.com/3524595/175629003-dcab5609-eead-42bb-8f34-12782ae501c8.png)

I have two different files, same code (except that the array key ) - on the left side psalm is happy.
on the right side, psalm is not. why ? 😂

**Comment by AndrolGenhald** [comment]
> can you make this work without any function?

$fqcn is [known to be exactly `'\\this\\is\\my\\test\\bla'`](https://psalm.dev/r/04d3cd8ad7), so you can just as easily [use the FQCN](https://psalm.dev/r/f2641879ab) with the `::class` constant.

> but if you see my screenshot, you might understand why this is totally confusing me

It's probably because you don't declare `$manHelperMapping`, Psalm doesn't like it when you do array assignment on an unset variable. [Here](https://psalm.dev/r/4d0b312bfc) it will assume the array might have other unknown values, so it's `mixed`, but [here](https://psalm.dev/r/22d3273c6f) it knows the only possible value is `"foobar"`.

**Comment by AndrolGenhald** [comment]
@orklah I'm fixing up a lot of this so it's consistent, but I do think we should prefer using `::class` instead of literals, would you be ok allowing [this](https://psalm.dev/r/4dd652fa39) and [this](https://psalm.dev/r/556784d485), but showing a new `LiteralStringAsClassname` (naming?) issue for [this](https://psalm.dev/r/f6bd100e94) and [this](https://psalm.dev/r/feb24cd951)?

**Comment by michabbb** [comment]
@AndrolGenhald to be sure, i replace "manhelpermapping" in my code with a fixed variable, same result:

left: OK
right: NOT OK

![image](https://user-images.githubusercontent.com/3524595/175661299-c96624c6-512b-454b-b6dd-3dfc3ab2a589.png)

I don´t really have an explanation for that 🤷

**Comment by AndrolGenhald** [comment]
@michabbb Add a `@psalm-trace` for `$HelperClassTexte` for the left one, maybe Psalm isn't analyzing the file or has bailed out early due to a previous issue?

**Comment by orklah** [comment]
@AndrolGenhald oh, those are weird errors indeed. Adding a type seems overkill though. Maybe allowing string literals in the list of known existing classes so we don't scream when we see one we already checked? (I didn't look at the code so I may be completely wrong here)

**Comment by AndrolGenhald** [comment]
@orklah We already have `Context::$phantom_classes` for that, it's just being ignored for string literals. I did some work on it but got sidetracked by string interpolation :laughing:

The real issue is that for this to work correctly I think we need to have context-specific type comparisons, because inside `if (class_exists("Foobar"))`, `"Foobar"` should be a subtype of `class-string`, but outside of the `if` it should not be.

**Comment by michabbb** [comment]
Here is some more info in case someone is interested. in my particular case (s. screenshots) I did some more tests.
the file you see on the right side is terribly old and has 2300 lines of code. first of all, I updated psalm because I was
using  `Psalm v4.15.0` and now `Psalm v4.23.0` - but the good thing is, that doesn´t change anything 😏 

next, I duplicated that terrible large file and reduced (nearly) every piece of code that is unrelated to this issue.
so I made this:

```php
<?php
declare(strict_types=1);

class Mode_db2ox2 extends Mode
{

    public function do_your_job(array $what = []): void
    {
        switch ($what) {
            case 'this':
                $text             = 'bs';
                $HelperClassTexte = '\\webtool\\getter\\lister\\template\\helper\\' . $text . '\\ebay\\v3\\TplHelperTexte';
                if (class_exists($HelperClassTexte)) {
                    echo "class exists!\n";
                }
                $this->test();
                break;
            default:
                echo "nothing todo!\n";
                break;
        }
    }

    public function test(array $what = []): array
    {
        $text             = 'bs';
        $HelperClassTexte = '\\webtool\\getter\\lister\\template\\helper\\' . $text . '\\ebay\\v3\\TplHelperTexte';
        if (class_exists($HelperClassTexte)) {
            echo "class exists!\n";
        }
        return [];
    }

}
```
which is in the case of a structure identical to that part inside the large file.  and maybe it´s worth mentioning that I never scanned the whole project because this takes minutes, I only scanned the particular file. but because both files in my example
extend the same class, I hope that this always has been a fair test, because the necessary scanned classes and files, should be the same. @AndrolGenhald so with that, I was always sure that psalm is scanning that file so I don´t need `@psalm-trace` here 😏 

TLDR; with the minified example, psalm is **NOT** complaining. so I still don´t know the reason, but a 2300 lines file seems to make any problems here. if one of the maintainers is willing to debug that, I am happy to help, but because I cannot share any more code here, we have to do this privately.
